plugins/com.google.cloud.tools.eclipse.appengine.libraries.ui/src/com/google/cloud/tools/eclipse/appengine/libraries/ui/CloudLibrariesPage.java

```diff
@@ -18,106 +18,197 @@
-import com.google.cloud.tools.eclipse.appengine.ui.AppEngineImages;
+import com.google.cloud.tools.eclipse.ui.util.images.SharedImages;
 import com.google.cloud.tools.eclipse.util.MavenUtils;
+import com.google.cloud.tools.eclipse.util.status.StatusUtil;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Maps;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 import org.eclipse.core.runtime.CoreException;
-import org.eclipse.core.runtime.SubMonitor;
+import org.eclipse.core.runtime.NullProgressMonitor;
 import org.eclipse.jdt.core.IClasspathEntry;
 import org.eclipse.jdt.core.IJavaProject;
 import org.eclipse.jdt.ui.wizards.IClasspathContainerPage;
 import org.eclipse.jdt.ui.wizards.IClasspathContainerPageExtension;
-import org.eclipse.jdt.ui.wizards.IClasspathContainerPageExtension2;
+import org.eclipse.jface.viewers.StructuredSelection;
 import org.eclipse.jface.wizard.WizardPage;
 import org.eclipse.swt.SWT;
-import org.eclipse.swt.layout.GridLayout;
+import org.eclipse.swt.layout.RowLayout;
 import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Group;
 import org.eclipse.wst.common.project.facet.core.IProjectFacetVersion;
 
-public abstract class CloudLibrariesPage extends WizardPage implements IClasspathContainerPage,
-    IClasspathContainerPageExtension, IClasspathContainerPageExtension2 {
-
+/**
+ * Provides for selecting from a suite of useful libraries for GCP projects. Supports editing the
+ * list of libraries for native Cloud Tools for Eclipse projects. Supported only selecting new
+ * libraries for Maven-based projects.
+ */
+public class CloudLibrariesPage extends WizardPage
+    implements IClasspathContainerPage, IClasspathContainerPageExtension {
   private static final Logger logger = Logger.getLogger(CloudLibrariesPage.class.getName());
-  private LibrarySelectorGroup librariesSelector;
+
+  /**
+   * The library groups to be displayed; pairs of (id, title). For example, <em>"clientapis" &rarr;
+   * "Google Cloud APIs for Java"</em>.
+   */
+  private Map<String, String> libraryGroups;
+
+  /** The initially selected libraries. */
+  private List<Library> initialSelection = Collections.emptyList();
+
+  @VisibleForTesting
+  final List<LibrarySelectorGroup> librariesSelectors = new ArrayList<>();
   private IJavaProject project;
-  private final String group;
-
-  protected CloudLibrariesPage(String group) {
-    super(group + "-page"); //$NON-NLS-1$
-    setTitle(Messages.getString(group + "-title"));  //$NON-NLS-1$
-    setDescription(Messages.getString(group + "-description"));  //$NON-NLS-1$
-    setImageDescriptor(AppEngineImages.appEngine(64));
-    this.group = group;
+  private boolean isMavenProject;
+  private IClasspathEntry oldEntry;
```

Comment:
`oldEntry` is assigned but never referenced. Remove it?

Reply:
Gone and good riddance.  I can't remember why I had it.